runtime/src/expected_rent_collection.rs

```diff
@@ -209,7 +209,115 @@ Every highest_root >= 432k * 2 + 80 and < 432k * 3 + 80 is this same result
+    /// returns Some(rent_epoch) if an adjustment needs to be made
+    /// returns None if the account is up to date
+    fn get_corrected_rent_epoch_on_load(
+        account: &AccountSharedData,
+        storage_slot: Slot,
+        bank_slot: Slot,
+        epoch_schedule: &EpochSchedule,
+        rent_collector: &RentCollector,
+        pubkey: &Pubkey,
+        rewrites_skipped_this_slot: &Rewrites,
+    ) -> Option<Epoch> {
+        if let RentResult::CollectRent((next_epoch, rent_due)) =
+            rent_collector.calculate_rent_result(pubkey, account, None)
+        {
+            if rent_due != 0 {
+                // rent is due on this account in this epoch, so we did not skip a rewrite
+                return None;
+            }
+
+            let (current_epoch, partition_from_current_slot) =
+                epoch_schedule.get_epoch_and_slot_index(bank_slot);
+            let (storage_epoch, storage_slot_partition) =
+                epoch_schedule.get_epoch_and_slot_index(storage_slot);
+            let partition_from_pubkey = Bank::partition_from_pubkey(
+                pubkey,
+                epoch_schedule.get_slots_in_epoch(current_epoch),
+            );
+            let mut possibly_update = true;
+            if current_epoch == storage_epoch {
+                // storage is in same epoch as bank
+                if partition_from_pubkey > partition_from_current_slot {
+                    // we haven't hit the slot's rent collection slot yet, and the storage was within this slot, so do not update
+                    possibly_update = false;
+                }
+            } else if current_epoch == storage_epoch + 1 {
+                // storage is in the previous epoch
+                if storage_slot_partition >= partition_from_pubkey
+                    && partition_from_pubkey > partition_from_current_slot
+                {
+                    // we did a rewrite in last epoch and we have not yet hit the rent collection slot in THIS epoch
+                    possibly_update = false;
+                }
+            } // if more than 1 epoch old, then we need to collect rent because we clearly skipped it.
+            let rent_epoch = account.rent_epoch();
+            if possibly_update
+                && rent_epoch == 0
+                && current_epoch > 1
+                && !rewrites_skipped_this_slot.contains_key(pubkey)
+            {
+                // we know we're done
+                return None;
+            }
+
+            // if an account was written >= its rent collection slot within the last epoch worth of slots, then we don't want to update it here
+            if possibly_update && rent_epoch < current_epoch {
+                // todo here - this needs to see if WE are the slot which should have done the rewrite for a prior slot when rent collection was due. so, we need to look at roots_original and ancestors...
```

Comment:
Does this todo still need to be addressed?

Reply:
no, that comment predates my use of 'rewrites_skipped_this_slot' and was left over. removed.